Implemented Save System

ButterLix opened the pull request:
This is a full implementation of a save system for CBT where the game is saved whenever the player hits a checkpoint with presents to save. This was made just to make the game not a complete one-shot experience where you have to play it entirely in one sitting

ButterLix committed:
First commit of implemented Save System

This is the first commit of this pull requests which features a fully functioning save system into Christmas Ball Torture

Iseeicy commented:
@ButterAleks formally saying - thank you so so much for this contrib, it's VERY much needed.

I haven't entirely reviewed, but after glossing through the changes I saw that a bunch of tiles were moved into autoload. Why?

ButterLix committed:
Fixed Tiles being placed in autoloads for some reason

ButterLix committed:
Got rid of some unnecessary stuff

There was just a bit of left over code from my testing that wasn't really needed

ButterAleks commented:
> I haven't entirely reviewed, but after glossing through the changes I saw that a bunch of tiles were moved into autoload. Why?

I just made some commits that should fix that weird issue and also get rid of some unnecessary changes. Not sure why that happened in the first place but it's definitely fixed now.

hollyfromthenexus commented on tiles/tileset3.tres:
Can you revert this change? I think this is leftover from when the tiles were accidentally moved to a different folder.

hollyfromthenexus commented on tiles/doors/yellow_present_door.tscn:
I think this change can be reverted too.

hollyfromthenexus commented on tiles/checkpoint/checkpoint.tscn:
I think this change can be reverted too

hollyfromthenexus commented on project.godot:
```diff
@@ -31,6 +32,7 @@ AreaTitleManager="*res://autoloads/area_title_manager.tscn"
 
 window/size/viewport_height=672
 window/size/resizable=false
+window/size/always_on_top=true
```
Let's revert this

ButterLix replied:
Must have accidentally left that one while testing...

hollyfromthenexus commented on project.godot:
```diff
@@ -44,7 +46,7 @@ export/convert_text_resources_to_binary=false
 [editor_plugins]
 
-enabled=PackedStringArray("res://addons/ggs/plugin.cfg", "res://addons/netengine5/net.bobbo.character/plugin.cfg", "res://addons/netengine5/net.bobbo.dialog-graph/plugin.cfg", "res://addons/netengine5/net.bobbo.interactable/plugin.cfg", "res://addons/netengine5/net.bobbo.item/plugin.cfg", "res://addons/netengine5/net.bobbo.knowledge/plugin.cfg", "res://addons/netengine5/net.bobbo.player-controlle ...
+enabled=PackedStringArray("res://addons/ggs/plugin.cfg")
```
Does the game work with all of the netengine plugins disabled??? I don't think I ever really use netengine so I kinda believe it.

ButterLix replied:
From what I could see yes, since it never really seemed to yell at me for disabling the plugins so loading them would just be unnecessary.

hollyfromthenexus commented on main_level.tscn:
I think we can revert this change (?)

hollyfromthenexus commented on collected_present.tscn:
This should be reverted

hollyfromthenexus commented on collected_present.gd:
What is this change?? Is it just whitespace?? For now I think we should revert this

hollyfromthenexus commented on Saved Text.tscn:
Do you mind making this scene follow the existing scene convention of snake_case?

so: `saved_text.tscn`

hollyfromthenexus commented on autoloads/save_manager.gd:
```diff
@@ -0,0 +1,174 @@
+
+func save_game():
+	# If we have a checkpoint to save to
+	if current_checkpoint != null:
```
IMO you should use a guard clause here, because this whole function is in the if statement!

ButterLix replied:
Fair enough, that would probably make it better on the cache, wasn't really thinking about that when writing this.

hollyfromthenexus commented on autoloads/save_manager.gd:
```diff
@@ -0,0 +1,174 @@
+
+# Check if the file exists and if it isn't empty
+func save_file_exists() -> bool:
+	if FileAccess.file_exists(save_file_path):
```
A guard clause would also be good here.

hollyfromthenexus commented on autoloads/save_manager.gd:
```diff
@@ -0,0 +1,174 @@
+	saved_text = get_node("/root/GameEntry/Saved Text/Text")
+	
+	# If we actually loaded any data
+	if load_position != Vector2.ZERO:
```
A guard clause would also be good here,

hollyfromthenexus reviewed (changes requested):
There's still some unnecessary changes in this PR it looks like. Let's clean those up, and then I'll take a closer look at the save manager! So far so good - I have some thoughts abt how we can clean up the actual save file format & simplify that code, but let's get the rest of the PR cleaned first!

Thanks so much for your continued effort @ButterAleks

ButterLix committed:
Added guard clauses and fixed IDs

For some reason Godot seems to update the ids of assets seemingly randomly when updating a scene which Github seems to constantly update about. Added guard clauses to some functions and fixed accidentally leaving the window on top on

ButterLix committed:
Actually fixed weird random ID changes

Last time it didn't work since Godot wasn't having it, but now hopefully it works

ButterLix committed:
Fixed whatever was going on with present_rule.gd

hollyfromthenexus reviewed (changes requested):
Currently we're serializing and de-serializing the save file partially manually, and partially through JSON. IMO this is fragile and a little hard to visually trace from a code standpoint.

I propose that we move over to a Resource-based solution. Let's create a new Resource class to act as a save file. This class would contain exports for the current load position and the collection of saved presents. Then, we can use Godot's [ResourceSaver](https://docs.godotengine.org/en/stable/classes/class_resourcesaver.html) class to write the resource to disk - and [ResourceLoader](https://docs.godotengine.org/en/stable/classes/class_resourceloader.html) to load it FROM disk.

This solution has some good advantages. It leverage's Godot's existing serialization for exportable variables - which is good for keeping our code simple and reliable. It allows us to structure the save data in a way that plays nice with type hinting. It also gives us a debugging benefit, because then save files could be easily created / edited in the Godot Inspector...!

ButterLix committed:
Changed Save System to use Resources instead of Text Files

ButterLix committed:
Fixed a gameplay issue and updated credits

I made it so that now whenever you activate a checkpoint it also saves the game instead of only when you collect presents as a quality of life feature. If I didn't do this then for example in the final section of the game if you were to quit and load then you would have to navigate through tensegrity again to get back (you wouldn't need to collect the presents again but it's still annoying). I also updated the credits to include me since I programmed the save system, and it also feels nice to think I helped play a part in making this game. Especially since I was ill during the main development of it.
